### DISCUSSION TITLE
Issue with Addon-Attach (Request being sent on every keystroke)

### DISCUSSION THREAD

**Discussion by AnanduP03** [sourcelair/xterm.js#4650] [Q&A] [answered]
I have been working on a project where i need to send the command through a websocket connection which i initially set up using some variable which will store the value entered and then when the user press enter it will check if the user entered value is clear or exit and if its not it will sent the data over the websocket connection. But I recently saw about attach addon and i implemented that and found that on each key i press the data is being sent over the websocket connection. I checked with the attach addons code and found the same being mentioned there as 
 this._disposables.push(terminal.onData(data => this._sendData(data)));
 
 So i want to know if there is a way i can make it work like when i press enter only the command should be sent over the websocket connection. It would be really helpful if there is a way i can customize the behavior when i press Enter. should i make changes to the addons code to make it work or is it possible to do it from my code?

**Comment by Tyriar** [comment, marked as answer]
You're referring to `ICANON` aka canonical or cooked mode, you need a pty to do on the "client side" where xterm.js is which isn't there because it's in a web browser. You can see https://github.com/microsoft/vscode/issues/135644 to understand the problem more. I don't think there's a way to do this currently, there are WIP projects like https://github.com/jerch/browser-fakepty but it's not a trivial thing to implement.

**Reply by AnanduP03** [reply]
hmm i guess I'll stick to the custom configuration I used on onData to send the request

let commandEntered = "";
    term.onData((data) => {
      if (data === "\r") {
        if (commandEntered === "clear") {
          term.clear();
        }else if(commandEntered==="exit"){
          term.writeln("\r\nWarning!!!")
          term.write("Your connection will be closed");
          term.writeln("\r\n");
          ws.send("");
        }
         else {
          ws.send(commandEntered);
        }
        commandEntered = "";
      } else if (data === "\x7F") {
        if (!commandEntered == "") {
          let line = term.buffer.active
            .getLine(term.buffer.active.cursorY)
            .translateToString()
            .trim();
          if (line) {
            line = line.slice(0, line.length - 1);
            commandEntered = commandEntered.slice(0, commandEntered.length - 1);
            term.write("\b \b");
          }
        }
      }
      else {
        commandEntered = commandEntered + data;
        term.write(data);
      }
    });


I'll use some array to store the commands entered by user and iterate over them to show the previous commands and for history the array can be printed as it is I guess.

Thanks for the response
